File: sturdr/nav/estimation.py
"""**navigation.py**

======  ============================================================================================
file    sturdr/nav/navigation.py 
brief   Satellite navigation techniques.
date    October 2024
refs    1. "Principles of GNSS, Inertial, and Multisensor Integrated Navigation Systems", 2nd Edition, 2013
            - Groves
        2. "Global Positioning System: Signals, Measurements, and Performance", 2nd Edition, 2006
            - Misra & Enge
======  ============================================================================================
"""

import numpy as np
from numba import njit

from sturdr.dsp.discriminator import DllVariance, FllVariance, PllVariance
from sturdr.nav.clock import GetNavClock
from sturdr.utils.coordinates import eci2ecef
from sturdr.utils.constants import GPS_L1CA_CARRIER_FREQ, GPS_L1CA_CODE_FREQ, LIGHT_SPEED, OMEGA_DOT

BETA = LIGHT_SPEED / GPS_L1CA_CODE_FREQ
LAMBDA = LIGHT_SPEED / GPS_L1CA_CARRIER_FREQ
OMEGA_IE_E = np.array(
    [
        [0.0, 0.0, 0.0],
        [0.0, 0.0, +OMEGA_DOT],
        [0.0, -OMEGA_DOT, 0.0]
    ], 
    dtype=np.double
)

@njit(cache=True, fastmath=True)
def LeastSquares(sv_pos: np.ndarray[np.double],
                 sv_vel: np.ndarray[np.double],
                 psr   : np.ndarray[np.double],
                 psrdot: np.ndarray[np.double],
                 CNo   : np.ndarray[np.double], 
                 x     : np.ndarray[np.double]):
    """
    Least Squares solver for GNSS position, velocity, and timing terms

    Parameters
    ----------
    sv_pos : np.ndarray[np.double]
        Satellite ECEF positions [m]
    sv_vel : np.ndarray[np.double]
        Satellite ECEF velocities [m/s]
    psr : np.ndarray[np.double]
        Pseudorange measurements [m]
    psrdot : np.ndarray[np.double]
        Pseudorange-rate measurements [m/s]
    R : np.ndarray[np.double], optional
        Measurement weighting matrix, by default None
    x : np.ndarray[np.double], optional
        Initial state estimate, by default np.zeros(8, dtype=np.double)

    Returns
    -------
    x : np.ndarray[np.double], optional
        Initial state estimate, by default np.zeros(8, dtype=np.double)
    """
    N1              = psr.size                              # half number of measurements
    N               = psr.size + psrdot.size                # number of measurements
    one             = np.ones(N1, dtype=np.double)          # column of ones
    H               = np.zeros((N, 8), dtype=np.double)     # measurement matrix (Jacobian)
    H[:N1, 6]       = one
    H[N1:, 7]       = one
    dz              = np.zeros(N, dtype=np.double)          # Measurement innovation
    R               = np.diag(np.concatenate((
                            BETA**2 * DllVariance(CNo, 0.02),
                            (LAMBDA/2/np.pi)**2 * FllVariance(CNo, 0.02)
                        )))

    for _ in range(20):
        for i in range(N1):
            # predict approximate range and account for earth's rotation (Groves 8.34)
            dr           = x[0:3] - sv_pos[i,:]
            r            = np.sqrt(dr @ dr)
            omegatau     = OMEGA_DOT * r / LIGHT_SPEED
            somegatau    = np.sin(omegatau)
            comegatau    = np.cos(omegatau)
            C_i_e        = np.array(
                                [
                                    [ comegatau, somegatau, 0.0],
                                    [-somegatau, comegatau, 0.0],
                                    [       0.0,       0.0, 1.0]
                                ], 
                                dtype=np.double
                           )
            
            # predict pseudorange (Groves 8.35, 9.165)
            dr           = x[0:3] - C_i_e @ sv_pos[i,:]
            r            = np.sqrt(dr @ dr) # np.sqrt(dr[0]**2 + dr[1]**2 + dr[2]**2)
            u            = dr / r
            pred_psr     = r + x[6]
            
            # predict pseudorange-rate (Groves 8.44, 9.165)
            dv           = (x[3:6] + OMEGA_IE_E @ x[0:3]) - C_i_e @ (sv_vel[i,:] + OMEGA_IE_E @ sv_pos[i,:])
            udot         = (dv * r**2 - dr * (dv @ dr)) / r**3
            pred_psrdot  = u @ dv + x[7]
            
            # setup the measurement matrix and innovation
            H[i, 0:3]    = u
            H[N1+i, 0:3] = udot
            H[N1+i, 3:6] = u
            dz[i]        = psr[i] - pred_psr
            dz[N1+i]     = psrdot[i] - pred_psrdot

        # weighted least squares solution
        P = np.linalg.inv(H.T @ R @ H)
        dx = P @ H.T @ R @ dz
        x += dx
        if np.sqrt(dx[0]**2 + dx[1]**2 + dx[2]**2) < 1e-6:
            break
    
    # print(np.array2string(H, max_line_width=80, precision=6))
    # print(np.array2string(dz, max_line_width=80, precision=6))
    return x, P


@njit(cache=True, fastmath=True)
def LeastSquaresPos(sv_pos: np.ndarray[np.double],
                    psr   : np.ndarray[np.double],
                    R     : np.ndarray[np.double], 
                    x     : np.ndarray[np.double]):
    N1              = psr.size                              # half number of measurements
    one             = np.ones(N1, dtype=np.double)          # column of ones
    H               = np.zeros((N1, 4), dtype=np.double)     # measurement matrix (Jacobian)
    H[:, 3]         = one
    dz              = np.zeros(N1, dtype=np.double)          # Measurement innovation

    for _ in range(20):
        for i in range(N1):
            # predict approximate range and account for earths rotation (Groves 8.34)
            dr           = x[0:3] - sv_pos[i,:]
            r            = np.sqrt(dr @ dr)
            omegatau     = OMEGA_DOT * r / LIGHT_SPEED
            somegatau    = np.sin(omegatau)
            comegatau    = np.cos(omegatau)
            C_i_e        = np.array(
                                [
                                    [ comegatau, somegatau, 0.0],
                                    [-somegatau, comegatau, 0.0],
                                    [       0.0,       0.0, 1.0]
                                ], 
                                dtype=np.double
                           )
            
            # predict pseudorange (Groves 8.35, 9.165)
            dr           = x[0:3] - C_i_e @ sv_pos[i,:]
            r            = np.sqrt(dr @ dr) # np.sqrt(dr[0]**2 + dr[1]**2 + dr[2]**2)
            u            = dr / r
            pred_psr     = r + x[3]
            
            # setup the measurement matrix and innovation
            H[i, 0:3]    = u
            dz[i]        = psr[i] - pred_psr

        # weighted least squares solution
        P = np.linalg.inv(H.T @ R @ H)
        dx = P @ H.T @ R @ dz
        x += dx
        if np.sqrt(dx[0]**2 + dx[1]**2 + dx[2]**2) < 1e-6:
            break

    return x, P, H


@njit(cache=True, fastmath=True)
def LeastSquaresVel(sv_vel: np.ndarray[np.double], 
                    psrdot: np.ndarray[np.double],
                    H     : np.ndarray[np.double],
                    R     : np.ndarray[np.double],
                    drift : np.double=0.0):
    if R is None:
        R = np.eye(psrdot.size)
        
    z = psrdot - (np.sum(H[:, :3] * -sv_vel, 1) + drift)
    P = np.linalg.inv(H.T @ R @ H)
    x = P @ H.T @ R @ z
    
    return x, P

class NavKF:
    """
    Kalman Filter for ECEF navigation states
    """

    __slots__ = 'x', 'P', 'A', 'Q', 'R', 'process_std', 'clock_model', 'innov_std', 'T'
    x           : np.ndarray[np.double]
    P           : np.ndarray[np.double]
    A           : np.ndarray[np.double]
    Q           : np.ndarray[np.double]
    process_std : np.double
    clock_model : str
    innov_std   : np.double
    T           : np.double
    
    def __init__(self, 
                 x: np.ndarray[np.double], 
                 P: np.ndarray[np.double], 
                 process_std: np.double, 
                 clock_model: str,
                 T: np.double):
        self.x = x
        self.P = P
        self.process_std = process_std
        self.clock_model = clock_model
        self.innov_std = 2.9
        self.T = T
        self.make_A(self.T)
        self.make_Q(self.T)
        return
    
    def run(self, 
            sv_pos: np.ndarray[np.double], 
            sv_vel: np.ndarray[np.double], 
            psr: np.ndarray[np.double], 
            psrdot: np.ndarray[np.double], 
            CNo: np.ndarray[np.double],
            T: np.double=None):

        self.x, self.P = self.Propagate(T)
        self.x, self.P, _,_ = self.Correct(sv_pos, sv_vel, psr, psrdot, CNo)
        return self.x, self.P
    
    def Propagate(self, T: np.double=None):
        if T is not None:
            self.make_A(T)
            self.make_Q(T)
        x = self.A @ self.x
        Q_2 = 0.5 * self.Q
        P = self.A @ (self.P + Q_2) @ self.A.T + Q_2
        return x, P
    
    def Correct(self, 
                sv_pos: np.ndarray[np.double], 
                sv_vel: np.ndarray[np.double], 
                psr: np.ndarray[np.double], 
                psrdot: np.ndarray[np.double], 
                CNo: np.ndarray[np.double]):
        # Correction
        N1              = psr.size                              # half number of measurements
        N               = psr.size + psrdot.size                # number of measurements
        one             = np.ones(N1, dtype=np.double)          # column of ones
        H               = np.zeros((N, 8), dtype=np.double)     # measurement matrix (Jacobian)
        H[:N1, 6]       = one
        H[N1:, 7]       = one
        dz              = np.zeros(N, dtype=np.double)          # Measurement innovation
        if N1 > 1:
            R               = np.diag(np.concatenate((
                                    BETA**2 * DllVariance(CNo, 0.02),
                                    (LAMBDA/2/np.pi)**2 * FllVariance(CNo, 0.02)
                            )))
            pred_psr        = np.zeros(N1)
            pred_psrdot     = np.zeros(N1)
            
            for i in range(N1):
                pred_psr[i], pred_psrdot[i], u, udot = PredictRangeAndRate(self.x, sv_pos[i,:], sv_vel[i,:])
                
                # setup the measurement matrix and innovation
                H[i, 0:3]    = u
                H[N1+i, 0:3] = udot
                H[N1+i, 3:6] = u
                dz[i]        = psr[i] - pred_psr[i]
                dz[N1+i]     = psrdot[i] - pred_psrdot[i]
        else:
            R = np.diag([BETA**2 * DllVariance(CNo, 0.02), (LAMBDA/2/np.pi)**2 * FllVariance(CNo, 0.02)])
            pred_psr, pred_psrdot, u, udot = PredictRangeAndRate(self.x, sv_pos, sv_vel)
            H[0, 0:3] = u
            H[1, 0:3] = udot
            H[1, 3:6] = u
            dz[0]     = psr - pred_psr
            dz[1]     = psrdot - pred_psrdot
            
        # innovation filter
        S = H @ self.P @ H.T + R
        norm_dz = np.abs(dz / np.sqrt(S.diagonal()))
        mask = norm_dz < self.innov_std
        dz = dz[mask]
        H = H[mask,:]
        R = np.diag(R.diagonal()[mask])
        
        # kalman update
        x = self.x
        P = self.P
        if dz.size > 0:
            PHt = P @ H.T
            K = PHt @ np.linalg.inv(H @ PHt + R)
            L = np.eye(8) - K @ H
            P = L @ P @ L.T + K @ R @ K.T
            x += K @ dz
        
        return x, P, pred_psr, pred_psrdot

    def make_A(self, T):
        self.A = np.asarray(
            [
                [1.0, 0.0, 0.0,   T, 0.0, 0.0, 0.0, 0.0],
                [0.0, 1.0, 0.0, 0.0,   T, 0.0, 0.0, 0.0],
                [0.0, 0.0, 1.0, 0.0, 0.0,   T, 0.0, 0.0],
                [0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0],
                [0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0],
                [0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0],
                [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0,   T],
                [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0],
            ],
            dtype=np.double
        )
        return
    
    def make_Q(self, T):
        av = GetNavClock(self.clock_model)
        q_bb = LIGHT_SPEED**2 * (av.h0/2*T + 2*av.h1*T**2 + (2/3)*np.pi**2*av.h2*T**3)
        q_dd = LIGHT_SPEED**2 * (av.h0/(2*T) + 2*av.h1 + (8/3)*np.pi**2*av.h2*T)
        q_bd = LIGHT_SPEED**2 * (2*av.h1*T + np.pi**2*av.h2*T**2)
        q_pp = self.process_std**2 * T**3 / 3.0
        q_pv = self.process_std**2 * T**2 / 2.0
        q_vv = self.process_std**2 * T
        self.Q = np.asarray(
            [
                [q_pp, 0.0, 0.0, q_pv, 0.0, 0.0, 0.0, 0.0],
                [0.0, q_pp, 0.0, 0.0, q_pv, 0.0, 0.0, 0.0],
                [0.0, 0.0, q_pp, 0.0, 0.0, q_pv, 0.0, 0.0],
                [q_pv, 0.0, 0.0, q_vv, 0.0, 0.0, 0.0, 0.0],
                [0.0, q_pv, 0.0, 0.0, q_vv, 0.0, 0.0, 0.0],
                [0.0, 0.0, q_pv, 0.0, 0.0, q_vv, 0.0, 0.0],
                [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, q_bb, q_bd],
                [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, q_bd, q_dd],
            ]
        )
        return
    
    
@njit(cache=True, fastmath=True)
def PredictRangeAndRate(x, sv_pos, sv_vel):
    # predict approximate range and account for earth's rotation (Groves 8.34)
    dr           = x[0:3] - sv_pos
    r            = np.sqrt(dr @ dr)
    omegatau     = OMEGA_DOT * r / LIGHT_SPEED
    somegatau    = np.sin(omegatau)
    comegatau    = np.cos(omegatau)
    C_i_e        = np.asarray(
                        [
                            [ comegatau, somegatau, 0.0],
                            [-somegatau, comegatau, 0.0],
                            [       0.0,       0.0, 1.0]
                        ], 
                        dtype=np.double
                    )
    
    # predict pseudorange (Groves 8.35, 9.165)
    dr           = x[0:3] - C_i_e @ sv_pos
    r            = np.sqrt(dr @ dr) # np.sqrt(dr[0]**2 + dr[1]**2 + dr[2]**2)
    u            = dr / r
    pred_psr     = r + x[6]
    
    # predict pseudorange-rate (Groves 8.44, 9.165)
    dv           = (x[3:6] + OMEGA_IE_E @ x[0:3]) - C_i_e @ (sv_vel + OMEGA_IE_E @ sv_pos)
    udot         = (dv * r**2 - dr * (dv @ dr)) / r**3
    pred_psrdot  = u @ dv + x[7]
    
    return pred_psr, pred_psrdot, u, udot
    

File: sturdr/sturdr/nav/estimation.cpp
/**
 * *estimation.cpp*
 *
 * =======  ========================================================================================
 * @file    sturdr/nav/estimation.cpp
 * @brief   Satellite navigation techniques.
 * @date    December 2024
 * @author  Daniel Sturdivant <sturdivant20@gmail.com>
 * @ref     1. "Principles of GNSS, Inertial, and Multisensor Integrated Navigation Systems", 2nd
 *              Edition, 2013 - Groves
 *          2. "Global Positioning System: Signals, Measurements, and Performance", 2nd Edition,
 *              2006 - Misra & Enge
 * =======  ========================================================================================
 */

#include "sturdr/nav/estimation.hpp"

#include <spdlog/spdlog.h>

#include <navtools/constants.hpp>

#include "sturdr/dsp/discriminator.hpp"

namespace sturdr {

inline static const Eigen::Matrix<double, 8, 8> I8 = Eigen::Matrix<double, 8, 8>::Identity();

void RangeAndRate(
    const Eigen::Ref<const Eigen::Vector3d> &pos,
    const Eigen::Ref<const Eigen::Vector3d> &vel,
    const double &cb,
    const double &cd,
    const Eigen::Ref<const Eigen::Vector3d> &sv_pos,
    const Eigen::Ref<const Eigen::Vector3d> &sv_vel,
    Eigen::Ref<Eigen::Vector3d> u,
    Eigen::Ref<Eigen::Vector3d> udot,
    double &pred_psr,
    double &pred_psrdot) {
  // predict approximate range and account for earth's rotation (Groves 8.34)
  Eigen::Vector3d dr = pos - sv_pos;
  double r = dr.norm();
  double wtau = navtools::WGS84_OMEGA<double> * r / navtools::LIGHT_SPEED<double>;
  double swtau = std::sin(wtau);
  double cwtau = std::cos(wtau);
  Eigen::Matrix3d C_i_e{{cwtau, swtau, 0.0}, {-swtau, cwtau, 0.0}, {0.0, 0.0, 1.0}};

  // predict pseudorange (Groves 8.35, 9.165)
  dr = pos - C_i_e * sv_pos;
  r = dr.norm();
  u = dr.array() / r;
  pred_psr = r + cb;

  // predict pseudorange-rate (Groves 8.44, 9.165)
  double r2 = r * r;
  Eigen::Vector3d dv = (vel + navtools::WGS84_OMEGA_SKEW<double> * pos) -
                       C_i_e * (sv_vel + navtools::WGS84_OMEGA_SKEW<double> * sv_pos);
  udot = (dv * r2 - dr * dv.dot(dr)) / (r2 * r);
  // std::cout << "dr = " << dr.transpose() << "\n";
  // std::cout << "dv = " << dv.transpose() << "\n";
  // std::cout << "dv * r^2 = " << (dv * r2).transpose() << "\n";
  // std::cout << "dv @ dr = " << dv.dot(dr) << "\n";
  // std::cout << "-dr * (dv @ dr) = " << (-dr * dv.dot(dr)).transpose() << "\n";
  pred_psrdot = u.dot(dv) + cd;
}

//! ------------------------------------------------------------------------------------------------

//* === NavKF ===
NavKF::NavKF(
    const Eigen::Vector<double, 8> &x,
    const Eigen::Matrix<double, 8, 8> &P,
    const double &T,
    const double &process_std,
    const double &innov_std,
    const std::string &clock_model)
    : x_{x},
      P_{P},
      clk_{GetNavClock(clock_model)},
      dyn_var_{0.5 * process_std * process_std},
      inn_std_{innov_std} {
  // Initialize dynamics matrices
  F_ = Eigen::Matrix<double, 8, 8>::Identity();
  Q_ = Eigen::Matrix<double, 8, 8>::Zero();
  make_F(T);
  make_Q(T);
}

// *=== ~NavKF ===*
NavKF::~NavKF() {
}

// *=== Dynamics ===*
bool NavKF::Dynamics() {
  try {
    x_ = F_ * x_;
    P_ = F_ * (P_ + Q_) * F_.transpose() + Q_;
    return true;
  } catch (std::exception &e) {
    spdlog::default_logger()->error("estimation.cpp NavKF::Dynamics failed! Error -> {}", e.what());
    return false;
  }
}
bool NavKF::Dynamics(const double &T) {
  make_F(T);
  make_Q(T);
  return Dynamics();
}

// *=== Measurement ===*
template <int N>
bool NavKF::Measurement(
    const Eigen::Matrix<double, N, 3> &sv_pos,
    const Eigen::Matrix<double, N, 3> &sv_vel,
    const Eigen::Vector<double, N> &psr,
    const Eigen::Vector<double, N> &psrdot,
    const Eigen::Vector<double, N> &cno,
    const double &beta,
    const double &lambda,
    const double &T) {
  try {
    // Initialize
    const int M = 2 * N;
    const Eigen::Vector<double, N> one = Eigen::Vector<double, N>::One();
    Eigen::Matrix<double, M, 8> H = Eigen::Matrix<double, M, 8>::Zero();
    H.col(6).segment<N>(0) = one;  // H.block<N, 1>(0, 6) = one;
    H.col(7).segment<N>(N) = one;  // H.block<N, 1>(N, 7) = one;
    Eigen::Vector<double, M> dz;

    double beta2 = beta * beta;
    double lambda2 = lambda * lambda;
    Eigen::Matrix<double, M, M> R = Eigen::Matrix<double, M, M>::Zero();
    for (int i = 0; i < M; i++) {
      R(i, i) = beta2 * DllVariance(cno(i), T);
      R(N + i, N + i) = lambda2 * FllVariance(cno(i), T);
    }

    // Innovation
    Eigen::Vector3d pos = x_.segment(0, 3);
    Eigen::Vector3d vel = x_.segment(3, 3);
    Eigen::Vector3d u, udot;
    double pred_psr, pred_psrdot;
    for (int i = 0; i < N; i++) {
      RangeAndRate(
          pos, vel, x_(6), x_(7), sv_pos.col(i), sv_vel.col(i), u, udot, pred_psr, pred_psrdot);

      // update the measurement matrix and innovation
      H.row(i).segment<3>(0) = u.transpose();         // H.block<1,3>(i,0)=u.transpose();
      H.row(N + i).segment<3>(0) = udot.transpose();  // H.block<1,3>(N+i,0)=udot.transpose();
      H.row(N + i).segment<3>(3) = udot.transpose();  // H.block<1,3>(N+i,3)=u.transpose();
      dz(i) = psr(i) - pred_psr;
      dz(N + i) = psrdot(i) - pred_psrdot;
    }

    // Innovation filter
    Eigen::Matrix<double, M, M> S = H * P_ * H.transpose() + R;
    Eigen::Vector<double, M> norm_dz = (dz.array() / S.diagonal().array().sqrt()).abs();
    Eigen::Array<bool, M, 1> mask = norm_dz.array() < inn_std_;
    const int M2 = mask.count();
    if (M2 < M) {
      // an innovation got flagged
      Eigen::Matrix<double, M2, 8> _H = Eigen::Matrix<double, M2, 8>::Zero();
      Eigen::Matrix<double, M2, M2> _R = Eigen::Matrix<double, M2, M2>::Zero();
      Eigen::Vector<double, M2> _dz;
      int j = 0;
      for (int i = 0; i < M; i++) {
        if (mask(i)) {
          _H.row(j) = H.row(i);
          _R(j, j) = R(i, i);
          _dz(j) = dz(i);
          j++;
        }
      }
      kalman_update(_H, _R, _dz);
    } else {
      // proceed as normal
      kalman_update(H, R, dz);
    }

    return true;
  } catch (std::exception &e) {
    spdlog::default_logger()->error(
        "estimation.cpp NavKF::Measurement failed! Error -> {}", e.what());
    return false;
  }
}

//* === make_F ===*
void NavKF::make_F(const double &T) {
  F_.block<3, 3>(0, 3).diagonal() << T, T, T;
  F_(6, 7) = T;
}

//* === make_Q ===*
void NavKF::make_Q(const double &T) {
  double LS2 = 0.5 * navtools::LIGHT_SPEED<double> * navtools::LIGHT_SPEED<double>;
  double T2 = T * T;
  double T3 = T * T2;
  double q_bb = LS2 * ((clk_.h0 * T / 2.0) + (2.0 * clk_.h1 * T2) +
                       (2.0 / 3.0 * navtools::PI_SQU<double> * clk_.h2 * T3));
  double q_bd = LS2 * ((2.0 * clk_.h1 * T) + (navtools::PI_SQU<double> * clk_.h2 * T2));
  double q_dd = LS2 * ((clk_.h0 / 2.0 / T) + (2.0 * clk_.h1) +
                       (8.0 / 3.0 * navtools::PI_SQU<double> * clk_.h2 * T));
  double q_pp = dyn_var_ * T3 / 3.0;
  double q_pv = dyn_var_ * T2 / 2.0;
  double q_vv = dyn_var_ * T;

  Q_.block<3, 3>(0, 0).diagonal() << q_pp, q_pp, q_pp;
  Q_.block<3, 3>(0, 3).diagonal() << q_pv, q_pv, q_pv;
  Q_.block<3, 3>(3, 0).diagonal() << q_pv, q_pv, q_pv;
  Q_.block<3, 3>(3, 3).diagonal() << q_vv, q_vv, q_vv;
  Q_(6, 6) = q_bb;
  Q_(6, 7) = q_bd;
  Q_(7, 6) = q_bd;
  Q_(7, 7) = q_dd;
}

// *=== kalman_update ===*
template <int M>
void NavKF::kalman_update(
    const Eigen::Matrix<double, M, 8> &H,
    const Eigen::Matrix<double, M, M> &R,
    const Eigen::Vector<double, 8> &dz) {
  Eigen::Matrix<double, 8, M> PHt = P_ * H.transpose();
  Eigen::Matrix<double, 8, M> K = PHt * (H * PHt + R).inverse();
  P_ = (I8 - K * H) * P_;
  x_ += K * dz;
}

}  // namespace sturdr